tensorflow/core/transforms/remapper/pass.cc

```diff
@@ -306,6 +312,241 @@ class BasePatternActivationRewriter : public BasePatternRewriter {
+        return false;
+      }
+    }
+
+    // Get the FusedBatchNorm training mode.
+    auto training_attr =
+        fused_batch_norm_op->getAttrOfType<BoolAttr>("is_training");
+    if (!training_attr) return false;
+    auto is_training = training_attr.getValue();
+
+    auto data_format_attr =
+        fused_batch_norm_op->getAttrOfType<StringAttr>("data_format");
+    if (!data_format_attr) return false;
+    auto data_format = data_format_attr.getValue();
+
+    if (data_format != "NHWC" && data_format != "NCHW") return false;
+
+    // In training mode we rely on cuDNN for computing FusedBatchNorm with side
+    // inputs and activation, and it has its own limitations. In inference mode
+    // we have a custom CUDA kernel that doesn't not have these constraints.
+    if (is_training &&
+        util::OpHasDevice(fused_batch_norm_op, tensorflow::DEVICE_GPU)) {
+      // cuDNN only supports NHWC data layout.
+      if (data_format != "NHWC") return false;
+
+      // Data type must be Float16.
+      if (!dtype_T.isa<Float16Type>()) return false;
+
+      // Channel dimension must be a multiple of 4.
+      auto fbn_input0_shape =
+          fused_batch_norm_op->getOperand(0).getType().cast<ShapedType>();
+      auto fbn_input0_shape_dims = fbn_input0_shape.getShape();
+
+      const bool valid_channel_dim = (fbn_input0_shape.getRank() == 4) &&
+                                     (fbn_input0_shape_dims[3] % 4 == 0);
+
+      if (!valid_channel_dim) return false;
+
+      // cuDNN must support CUDNN_BATCHNORM_SPATIAL_PERSISTENT mode.
+      if (!BatchnormSpatialPersistentEnabled()) return false;
+    }
+
+    // FusedBatchNormV2 and V3 have an extra type parameter.
+    if (fused_batch_norm_op->getName().getStringRef().str() !=
+        "tfg.FusedBatchNorm") {
+      auto attr = fused_batch_norm_op->getAttrOfType<TypeAttr>("U");
+      if (attr && !attr.getValue().isa<Float32Type>()) {
+        return false;
+      }
+    }
+
+    // Check that only one node consumes the 0-th output of a FusedBatchNorm.
+    if (this->helper_.HasControlOperandsOrResultUsers(fused_batch_norm_op) ||
+        !this->helper_.HasAtMostOneUserOfResult0(fused_batch_norm_op))
+      return false;
+
+    return true;
+  }
+
+  bool matchPattern(Operation *op, Pattern &pattern) const {
+    TFOp activation_tfg_wrapper(op);
+    // Not allowing control flow on Relu
+    if (helper_.HasControlOperandsOrResultUsers(op)) return false;
+    if (activation_tfg_wrapper.getNonControlOperands().empty()) return false;
+
+    Operation *activation_input_op = op->getOperand(0).getDefiningOp();
+    if (is_valid_batch_norm(activation_input_op)) {
+      pattern.fused_batch_norm = activation_input_op;
+      pattern.activation = op;
+      pattern.side_input = nullptr;
+      return true;
+    }
+
+    // Input to a Relu can be an Add node with FusedBatchNorm as one of the
+    // inputs
+    if (this->helper_.getDialect()->IsAdd(activation_input_op)) {
```

Comment:
Nit: Add brackets here and below two if clauses

Reply:
Done